### PR TITLE
Add v2 REST API specification

### PR DESCRIPTION
**Pull request by iain-anderson**
Signed-off-by: Iain Anderson <iain@iotechsys.com>

## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://wiki.edgexfoundry.org/display/FA/Committing+Code+Guidelines#CommittingCodeGuidelines-CommitMessages
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ x] Documentation content changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number:

## Does this PR introduce a breaking change?

- [ ] Yes
- [ ] No

## Other information

### CODE REVIEW

**Comment by iain-anderson** [comment]
Note in particular:

- `device/all` is removed
- `callback` is replaced by:

- - `callback/device` (post/put)
- - `callback/device/id` (delete)
- - `callback/profile` (post/put)
- - `callback/profile/id` (delete)
- - `callback/watcher` (post/put)
- - `callback/watcher/id` (delete)

**Comment by jpwhitemn** [comment]
LGTM Iain.  It looks like it shares much with the other OpenAPI docs including and most importantly command service.  One question, in this API spec, CommandAction items ref ActionResponse, whereas in command service API spec it ref's CommandResponse.  Was there a reason for this difference?

**Comment by iain-anderson** [comment]
Hi Jim, it's because the core-metadata and core-command API specs name them differently :)
This inconsistency has gone away now that edgex-go #2447 has landed

**Comment by iain-anderson** [comment]
Updated in line with https://github.com/edgexfoundry/edgex-go/pull/2447

File: api/oas3.0/v2/device-sdk.yaml [hunk 1]
```diff
@@ -0,0 +1,1269 @@
...
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/SettingRequest'
+        required: true
+
+  /discovery:
+    post:
+      description: Run the discovery request for a Device Service.
+      responses:
```
Comment (lenny-intel) [inline api/oas3.0/v2/device-sdk.yaml]: RequestBody? i.e. for consistency we should have a request DTO for every API? If not, then document that not request body required.

File: api/oas3.0/v2/device-sdk.yaml [hunk 2]
```diff
@@ -0,0 +1,1269 @@
...
+      allOf:
+        - $ref: '#/components/schemas/BaseRequest'
+      type: object
+    VersionResponse:
+      description: "A response returned from the /version endpoint whose purpose is to report out the latest version supported by the service."
+      allOf:
+      - $ref: '#/components/schemas/BaseResponse'
+      type: object
+      properties:
+        version:
```
Comment (lenny-intel) [inline api/oas3.0/v2/device-sdk.yaml]: For V1 the SDKs  are returning SDK version and Service version. Need to do the same for V2?

File: api/oas3.0/v2/device-sdk.yaml [hunk 3]
```diff
@@ -0,0 +1,1269 @@
...
+  version: 2.x
+
+servers:
+  - url: http://0:49999/api/v2
+    description: URL for local development and testing
+
+components:
+  schemas:
+    AutoEvent:
+      type: object
```
[the remark this answers is not in the document]
Comment (iain-anderson) [reply api/oas3.0/v2/device-sdk.yaml]: done

File: api/oas3.0/v2/device-sdk.yaml [hunk 4]
```diff
@@ -0,0 +1,1269 @@
...
+      required:
+        - requestId
+    NewBinaryReading:
+      description: "An event reading for a binary data type"
+      allOf:
+        - $ref: '#/components/schemas/NewBaseReading'
+        - type: object
+          properties:
+            binaryValue:
+              description: "If the value of the reading is binary, it will be found in this property as a byte array"
```
[the remark this answers is not in the document]
Comment (iain-anderson) [reply api/oas3.0/v2/device-sdk.yaml]: done

File: api/oas3.0/v2/device-sdk.yaml [hunk 5]
```diff
@@ -0,0 +1,1269 @@
...
+              type: string
```
[the remark this answers is not in the document]
Comment (iain-anderson) [reply api/oas3.0/v2/device-sdk.yaml]: Changed to format: binary

File: api/oas3.0/v2/device-sdk.yaml [hunk 6]
```diff
@@ -0,0 +1,1269 @@
...
+        name:
+          type: string
+        tag:
+          type: string
+        properties:
+          $ref: '#/components/schemas/PropertyValue'
+        attributes:
+          type: object
+          additionalProperties:
+            type: string
```
[the remark this answers is not in the document]
Comment (iain-anderson) [reply api/oas3.0/v2/device-sdk.yaml]: I don't think so. These are fields with arbitrary names.

File: api/oas3.0/v2/device-sdk.yaml [hunk 7]
```diff
@@ -0,0 +1,1269 @@
...
+      type: object
+    ConfigResponse:
+      allOf:
+        - $ref: '#/components/schemas/BaseResponse'
+      description: "Provides a response containing the configuration for the targeted service."
+      type: object
+      properties:
+        config:
+          description: "A string-ified representation of the service's configuration. For purposes of this specification, a string has been used since configuration structure differs from service to service."
+          type: string
```
[the remark this answers is not in the document]
Comment (iain-anderson) [reply api/oas3.0/v2/device-sdk.yaml]: done

File: api/oas3.0/v2/device-sdk.yaml [hunk 8]
```diff
@@ -0,0 +1,1269 @@
...
+      properties:
+        device:
+          $ref: '#/components/schemas/Device'
+    NewProfileRequest:
+      allOf:
+        - $ref: '#/components/schemas/BaseRequest'
+      description: "Notification that a new device profile has been created."
+      type: object
+      properties:
+        device:
```
[the remark this answers is not in the document]
Comment (iain-anderson) [reply api/oas3.0/v2/device-sdk.yaml]: Yes. copy-paste.

File: api/oas3.0/v2/device-sdk.yaml [hunk 9]
```diff
@@ -0,0 +1,1269 @@
...
+          $ref: '#/components/schemas/DeviceProfile'
+    NewProvisionWatcherRequest:
+      allOf:
+        - $ref: '#/components/schemas/BaseRequest'
+      description: "Notification that a new provision watcher associated with this device service has been created."
+      type: object
+      properties:
+        device:
```
[the remark this answers is not in the document]
Comment (iain-anderson) [reply api/oas3.0/v2/device-sdk.yaml]: fixed

File: api/oas3.0/v2/device-sdk.yaml [hunk 10]
```diff
@@ -0,0 +1,1269 @@
...
+        assertion:
+          type: string
+        floatEncoding:
+          type: string
+        mediaType:
+          type: string
+    ProtocolProperties:
+      type: object
+      additionalProperties:
+        type: string
```
[the remark this answers is not in the document]
Comment (iain-anderson) [reply api/oas3.0/v2/device-sdk.yaml]: As above, don't think so

File: api/oas3.0/v2/device-sdk.yaml [hunk 11]
```diff
@@ -0,0 +1,1269 @@
...
+          format: uuid
+        name:
+          type: string
+        created:
+          type: integer
+        modified:
+          type: integer
+        identifiers:
+          type: object
+          additionalProperties:
```
[the remark this answers is not in the document]
Comment (iain-anderson) [reply api/oas3.0/v2/device-sdk.yaml]: As above, don't think so

File: api/oas3.0/v2/device-sdk.yaml [hunk 12]
```diff
@@ -0,0 +1,1269 @@
...
+      properties:
+        device:
+          $ref: '#/components/schemas/Device'
+    UpdateProfileRequest:
+      allOf:
+        - $ref: '#/components/schemas/BaseRequest'
+      description: "Notification that an existing device profile has been updated."
+      type: object
+      properties:
+        device:
```
[the remark this answers is not in the document]
Comment (iain-anderson) [reply api/oas3.0/v2/device-sdk.yaml]: done

File: api/oas3.0/v2/device-sdk.yaml [hunk 13]
```diff
@@ -0,0 +1,1269 @@
...
+          $ref: '#/components/schemas/DeviceProfile'
+    UpdateProvisionWatcherRequest:
+      allOf:
+        - $ref: '#/components/schemas/BaseRequest'
+      description: "Notification that an existing provision watcher associated with this device service has been updated."
+      type: object
+      properties:
+        device:
```
[the remark this answers is not in the document]
Comment (iain-anderson) [reply api/oas3.0/v2/device-sdk.yaml]: done

File: api/oas3.0/v2/device-sdk.yaml [hunk 14]
```diff
@@ -0,0 +1,1269 @@
...
+          description: "OK"
+          headers:
+            X-Correlation-ID:
+              $ref: '#/components/headers/correlatedResponseHeader'
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/ConfigResponse'
+        '207':
+          description: "Indicates a multi-part response supportive of accepting multiple requests at once. The 'statusCode' property of each response in the returned array will indicate success or failure."
```
[the remark this answers is not in the document]
Comment (iain-anderson) [reply api/oas3.0/v2/device-sdk.yaml]: Don't know. Spec is taken from edgex-go. @tsconn23 ?

File: api/oas3.0/v2/device-sdk.yaml (hunk 1 above)
Comment (iain-anderson) [reply api/oas3.0/v2/device-sdk.yaml]: Noted no request body

File: api/oas3.0/v2/device-sdk.yaml (hunk 2 above)
Comment (iain-anderson) [reply api/oas3.0/v2/device-sdk.yaml]: Yes, amended

**Comment by iain-anderson** [comment]
Added descriptions. Added some "requires" specifications but many of these are context-dependent (eg the required fields for posting a new object are different to an update), am guessing this is why they aren't specified in edgex-go?

File: api/oas3.0/v2/device-sdk.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (lenny-intel) [reply api/oas3.0/v2/device-sdk.yaml]: `additionalProperties` indicates it is a list of some type. CSV?

[the remark this answers is not in the document]
Comment (iain-anderson) [reply api/oas3.0/v2/device-sdk.yaml]: I think it works as a dictionary as highlighted here: https://swagger.io/docs/specification/data-models/dictionaries/

**Comment by iain-anderson** [comment]
> Should it be our intent to have description for every object and every property?
The definitions in question are imported from edgex-go/api/openapi/v2 so I don't believe it's currently the intention. Perhaps it should be though.

**Comment by iain-anderson** [comment]
Added additional description fields

File: api/oas3.0/v2/device-sdk.yaml (hunk 14 above)
[the remark this answers is not in the document]
Comment (lenny-intel) [reply api/oas3.0/v2/device-sdk.yaml]: @tsconn23, Does this need to have a multi-part response?

[the remark this answers is not in the document]
Comment (tsconn23) [reply api/oas3.0/v2/device-sdk.yaml]: Multi-part is recommended to establish a common pattern for all request/responses. In practice the endpoint would return an array with only one element, but it's better in my view to do that than have one (or a handful) endpoint with a different behavior.

[the remark this answers is not in the document]
Comment (lenny-intel) [reply api/oas3.0/v2/device-sdk.yaml]: Got it. Thanks!

**Pull request merged**